Use github wf to add module labels for PR based on file changes

pseudo-nymous opened the pull request:
### Description
Adds a new bot to automate labelling of pull requests with module tags based on file changes.
This bot uses `.github/labeler.yml` file to maintain mapping for module labels to corresponding regex paths to track. 

This workflow is triggered using `pull_request_target` event which provides necessary write permission for adding labels to PRs. More details around this event: [StackOverflow](https://stackoverflow.com/questions/74957218/what-is-the-difference-between-pull-request-and-pull-request-target-event-in-git)

### Testing
Tested changes by creating pr on my own fork. https://github.com/pseudo-nymous/lucene/pull/2

Further testing with sync flag: https://github.com/pseudo-nymous/lucene/pull/3

Closes #13897

pseudo-nymous committed:
Use github wf to add module labels for PR based on file changes

stefanvodita commented:
@pseudo-nymous, I'm only seeing this now, sorry! At first glance, it matches what I had in mind - thank you for addressing that issue! I'll do an in-depth review soon, but I'd appreciate others' thoughts as well. One potential concern is our GitHub action budget - do you have a sense of the cost of this workflow?

stefanvodita commented on .github/labeler.yml:
```diff
@@ -0,0 +1,134 @@
+
+module:spatial-extras:
+  - changed-files:
+      - any-glob-to-any-file: 'lucene/spatial-extras/**'
```
Curious - why give this its own label and the `module:spatial` label as well?

pseudo-nymous replied:
It was a miss. I'll remove `lucene/spatial-extras/**` from `module:spatial` label.

stefanvodita replied:
Sorry, I should have made a recommendation. I think it would be better to list both paths under `module:spatial` and not have `module:spatial-extras` at all. `spatial-extras` is really the core spatial module ([overview](https://github.com/apache/lucene/blob/main/lucene/spatial-extras/src/java/overview.html)).

pseudo-nymous replied:
I see. It makes sense to have it in `module:spatial` label itself.

pseudo-nymous replied:
Updated.

stefanvodita commented on .github/labeler.yml:
```diff
@@ -0,0 +1,134 @@
+# This configuration is used by the workflow defined in .github/workflows/label-pull-request.yml.
+# If we are adding new labels or refactoring modules, we will need to modify this file globs here to ensure that the correct labels are applied.
+
+# For more information on how to use the `actions/checkout` action, visit:
```
Is this relevant? I don't see us using `checkout` here.

pseudo-nymous replied:
It's not relevant anymore. I'll remove it.

stefanvodita commented on .github/workflows/label-pull-request.yml:
```diff
@@ -0,0 +1,21 @@
+  - pull_request_target
+
+jobs:
+  labeler:
```
Can we add `sync-labels` to make it so labels get removed when they are no longer valid after an update to the PR?

pseudo-nymous replied:
Great suggestion! I'll update.

stefanvodita commented on .github/labeler.yml:
```diff
@@ -0,0 +1,134 @@
+module:core/other:
+  - changed-files:
+      - any-glob-to-any-file: ['lucene/core/**']
+      - all-globs-to-all-files:
```
Cool!

stefanvodita reviewed (approved):
This is great @pseudo-nymous! I left minor comments that I would like to see addressed, but this is very good already. I retract my concern about the action budget. This is running when a PR is created or updated, which is not that often.

pseudo-nymous commented:
Did further testing with new sync flag, change is working as expected.
Ref: https://github.com/pseudo-nymous/lucene/pull/3

pseudo-nymous commented:
> This is great @pseudo-nymous! I left minor comments that I would like to see addressed, but this is very good already. I retract my concern about the action budget. This is running when a PR is created or updated, which is not that often.

Thanks. 

Yes, this bot is only triggered when when pr is created or updated.

I have addressed all the comments. Please take a look.

stefanvodita reviewed (approved):
Thank you for responding to the comments so quickly! I've one more small thing to add and if that's sorted I'll leave it a few days to see if others have feedback, then merge.

pseudo-nymous committed:
Addressed comments and fixed glob for core module

mikemccand commented:
Wow, this looks really awesome!  I wish we could retroactively apply it to PRs missing their module labels ... but let's start with this change (so new PRs are labeled).

stefanvodita commented:
Thank you for this contribution @pseudo-nymous, I merged it just now and I'm really looking forward to see the first PR using it!

stefanvodita commented:
Brilliant! Let us know if you find anything. If you're interesting in this area of GitHub workflows / community tools, there are a few others that would be great to do:
1. #13898
2. #14190
3. #13896

pseudo-nymous commented:
I'll be happy to work on these. Thanks!

stefanvodita commented:
Linking the PR related to the question above: #14721.
